File: square-commons/src/main/java/com/raxdenstudios/square/interceptor/commons/injectfragmentlist/InjectFragmentListActivityInterceptor.java
package com.raxdenstudios.square.interceptor.commons.injectfragmentlist;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.view.View;

import com.raxdenstudios.square.interceptor.ActivityInterceptor;
import com.raxdenstudios.square.utils.FragmentUtils;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Ángel Gómez on 20/12/2016.
 */
public class InjectFragmentListActivityInterceptor<TFragment extends Fragment> extends ActivityInterceptor<InjectFragmentListInterceptorCallback<TFragment>> implements InjectFragmentListInterceptor {

    List<TFragment> mFragmentList;

    public InjectFragmentListActivityInterceptor(@NonNull FragmentActivity activity) {
        super(activity);
    }

    public InjectFragmentListActivityInterceptor(@NonNull FragmentActivity activity, @NonNull InjectFragmentListInterceptorCallback<TFragment> callback) {
        super(activity, callback);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (mFragmentList != null && !mFragmentList.isEmpty()) {
            for (TFragment fragment : mFragmentList) {
                fragment.onActivityResult(requestCode, resultCode, data);
            }
        }
    }

    @Override
    @SuppressWarnings("unchecked")
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        for (int i = 0; i < getCallback().getFragmentCount(); i++) {
            View contentView = getCallback().onLoadFragmentContainer(savedInstanceState, i);
            TFragment contentFragment;
            if (savedInstanceState == null) {
                contentFragment = getCallback().onCreateFragment(i);
                FragmentUtils.INSTANCE.loadFragment(getActivity().getSupportFragmentManager(), contentView.getId(), contentFragment);
            } else {
                contentFragment = (TFragment) FragmentUtils.INSTANCE.getFragment(getActivity().getSupportFragmentManager(), contentView.getId());
            }
            getCallback().onFragmentLoaded(contentFragment, i);
            addFragment(contentFragment);
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mFragmentList.clear();
        mFragmentList = null;
    }

    private void addFragment(TFragment fragment) {
        if (mFragmentList == null) mFragmentList = new ArrayList<>();
        mFragmentList.add(fragment);
    }

}

File: square-commons/src/main/java/com/raxdenstudios/square/interceptor/commons/injectfragmentlist/InjectFragmentListActivityInterceptor.kt
package com.raxdenstudios.square.interceptor.commons.injectfragmentlist

import android.app.Activity
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentActivity
import android.view.View
import com.raxdenstudios.square.interceptor.ActivityInterceptor

/**
 * Created by Ángel Gómez on 22/05/2015.
 */
class InjectFragmentListActivityInterceptor<TFragment : Fragment>(
        callback: HasInjectFragmentListInterceptor<TFragment>
) : ActivityInterceptor<InjectFragmentListInterceptor, HasInjectFragmentListInterceptor<TFragment>>(callback),
        InjectFragmentListInterceptor {

    private var mContainerViewMap: MutableMap<Int, View> = mutableMapOf()
    private var mContainerFragmentMap: MutableMap<Int, TFragment?> = mutableMapOf()

    override fun onActivityCreated(activity: Activity?, savedInstanceState: Bundle?) {
        super.onActivityCreated(activity, savedInstanceState)

        getFragmentManager(activity)?.also { fm ->
            if (savedInstanceState == null) {
                for (position in 0 until mCallback.fragmentCount) {
                    mContainerViewMap[position] = mCallback.onLoadFragmentContainer(position).also { view ->
                        mContainerFragmentMap[position] = mCallback.onCreateFragment(position).also {
                            fm.beginTransaction()
                                    .replace(view.id, it, it.javaClass.simpleName)
                                    .commit()
                            mCallback.onFragmentLoaded(position, it)
                        }
                    }
                }
            }
        }
    }

    override fun onActivityStarted(activity: Activity?) {
        super.onActivityStarted(activity)

        getFragmentManager(activity)?.also { fm ->
            if (mSavedInstanceState != null) {
                for (position in 0 until mCallback.fragmentCount) {
                    mContainerViewMap[position]?.also { view ->
                        mContainerFragmentMap[position] = (fm.findFragmentById(view.id) as? TFragment)?.also {
                            mCallback.onFragmentLoaded(position, it)
                        }
                    }
                }
            }
        }
    }

    override fun onActivityDestroyed(activity: Activity?) {
        super.onActivityDestroyed(activity)

        mContainerViewMap.clear()
        mContainerFragmentMap.clear()
    }
}

